rotkehlchen/exchanges/woo.py

```diff
@@ -0,0 +1,431 @@
+            except JSONDecodeError:
+                msg = f'{self.name} {case} returned an invalid JSON response: {response.text}.'
+                log.error(msg, options=call_options)
+                self.msg_aggregator.add_error(
+                    f'Got remote error while querying {self.name} {case}: {msg}',
+                )
+                return []
+            deserialization_method = self._deserialize_trade
+            entries = response_data['data']
+
+            for entry in entries:
+                try:
+                    result = deserialization_method(entry)
+                    results.append(result)
+                except DeserializationError as e:
+                    log.error(
+                        'Error processing a Woo balance.',
+                        entry=entry,
+                        error=str(e),
+                    )
+            if len(entries) < call_options['limit']:
+                break
+            call_options['fromId'] = entries[-1]['id']
+
+        return results
+
+    @overload
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: Literal['validate_api_key'],
+    ) -> tuple[Literal[False], str]:
+        ...
+
+    @overload
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: Literal['balances'],
+    ) -> ExchangeQueryBalances:
+        ...
+
+    @overload
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: Literal['trades'],
+    ) -> list[Trade]:
+        ...
+
+    @overload
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> list[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
+            self,
+            response: Response,
+            case: Literal['validate_api_key', 'balances', 'trades', 'asset_movements'],
+    ) -> Union[
+        list,
+        tuple[Literal[False], str],
+        ExchangeQueryBalances,
+    ]:
+        """This function processes not successful responses for the following
+        cases listed in `case`.
+        """
+        case_pretty = case.replace('_', ' ')  # human readable case
+        try:
+            response_dict = jsonloads_dict(response.text)
+        except JSONDecodeError as e:
+            msg = f'{self.name} returned invalid JSON response: {response.text}.'
+            log.error(msg)
+
+            if case in ('validate_api_key', 'balances'):
+                return False, msg
+            if case in ('trades', 'asset_movements'):
+                self.msg_aggregator.add_error(
+                    f'Got remote error while querying {self.name} {case}: {msg}',
+                )
+                return []
+
+            raise AssertionError(f'Unexpected {self.name} response_case: {case_pretty}.') from e
```

Comment:
don't raise this kind of errors in the code. Does are programming errors. You can add a simple assert that will be removed when packaging the app